components/org.wso2.carbon.identity.application.authenticator.sessionauth/src/main/java/org/wso2/carbon/identity/application/authenticator/sessionauth/SessionCountAuthenticator.java

```diff
@@ -0,0 +1,243 @@
+import java.util.Map;
+import java.util.Set;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import static java.lang.Integer.parseInt;
+import static org.wso2.carbon.identity.application.authenticator.sessionauth.util.AuthenticatorUtil.getSessionDetails;
+
+/**
+ * Session count based authenticator.
+ */
+public class SessionCountAuthenticator extends AbstractApplicationAuthenticator
+        implements LocalApplicationAuthenticator {
+
+    private static final long serialVersionUID = 1819664539416029245L;
+    private static final Log log = LogFactory.getLog(SessionCountAuthenticator.class);
+
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+
+        return true;
+    }
+
+    @Override
+    public AuthenticatorFlowStatus process(HttpServletRequest request,
+                                           HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException {
+        if (context.isLogoutRequest()) {
+            return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+        }
+        //checks whether request has gone through the session.jsp page or not
+        if (StringUtils.isNotEmpty(request.getParameter(
+                SessionCountAuthenticatorConstants.SESSION_TERMINATION_SERVLET_INPUT))) {
+            try {
+                processAuthenticationResponse(request, response, context);
+            } catch (SessionValidationException e) {
+                context.setRetrying(true);
+                context.setCurrentAuthenticator(getName());
+                return initiateAuthRequest(response, context,
+                        "Exception occurred in session termination process. Please try again");
+            }
+            return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+        } else {
+            return initiateAuthRequest(response, context, null);
+        }
+    }
+
+    /**
+     * This will gather information on exiting sessions of the current user trying to execute the authenticator and
+     * store them as metadata and redirect user to page to select sessions to termination.
+     *
+     * @param response     the response
+     * @param context      the authentication context
+     * @param errorMessage contains error message of previous attempt if this is a retry attempt.
+     * @throws AuthenticationFailedException when a problem occurred in session termination process, user
+     *                                       authentication process or meta data retrieving process
+     */
+    private AuthenticatorFlowStatus initiateAuthRequest(HttpServletResponse response, AuthenticationContext context,
+                                                        String errorMessage) throws AuthenticationFailedException {
+
+        //Identifying the authenticated user from the previous step
+        StepConfig stepConfig = context.getSequenceConfig().getStepMap().get(context.getCurrentStep() - 1);
```

Comment:
This would give NPE in scenarios where this is configured as the first step (accidentally may be). Please handle that.

Reply:
fixed in the new commit